mlir/lib/Catalyst/Transforms/DetensorizeSCFPass.cpp

```diff
@@ -0,0 +1,354 @@
+            rewriter.setInsertionPointAfter(forOp);
+            for (std::size_t index : newIterOperandsIndices) {
+                Value for_result = newForOp->getResult(index);
+                Value value = rewriter.create<tensor::FromElementsOp>(
+                    newForOp->getLoc(), RankedTensorType::get({}, for_result.getType()),
+                    for_result);
+                rewriter.replaceUsesWithIf(forOp->getResult(index), value, [&](OpOperand &op) {
+                    return !isa<tensor::FromElementsOp>(op.getOwner());
+                });
+            }
+        }
+
+        rewriter.replaceOp(forOp, newForOp);
+        return success();
+    }
+};
+
+struct DetensorizeIfOp : public OpRewritePattern<scf::IfOp> {
+    using OpRewritePattern<scf::IfOp>::OpRewritePattern;
+
+    LogicalResult matchAndRewrite(scf::IfOp ifOp, PatternRewriter &rewriter) const override
+    {
+        // Early exit if there are no results that could be replaced.
+        if (ifOp.getNumResults() == 0) {
+            return failure();
+        }
+        if (!hasScalarTensorResult(ifOp)) {
+            return failure();
+        }
+
+        // 1. Extract tensor elements before yield op
+        SmallVector<scf::YieldOp> yieldOps = {ifOp.thenYield(), ifOp.elseYield()};
+        for (scf::YieldOp yield_op : yieldOps) {
+            // Loop over yield operands
+            for (const auto &it : llvm::enumerate(yield_op.getOperands())) {
+                Value operand = it.value();
+                // Detensorize operand: extract tensor element before yielding
+                if (isScalarTensor(operand)) {
+                    OpBuilder::InsertionGuard g(rewriter);
+                    rewriter.setInsertionPoint(yield_op);
+                    Value value = rewriter.create<tensor::ExtractOp>(yield_op->getLoc(), operand,
+                                                                     ValueRange{});
+                    yield_op.setOperand(it.index(), value);
+                }
+            }
+        };
+
+        // Collect the types for the new IfOp
+        SmallVector<Type> newResultTypes;
+        for (auto result : ifOp->getResults()) {
+            if (isScalarTensor(result)) {
+                newResultTypes.push_back(cast<TensorType>(result.getType()).getElementType());
+                continue;
+            }
+            newResultTypes.push_back(result.getType());
+        }
+
+        // 2. Create the new IfOp
+        OpBuilder::InsertionGuard ifOpInsertionGuard(rewriter);
+        rewriter.setInsertionPoint(ifOp);
+        auto newIfOp =
+            rewriter.create<scf::IfOp>(ifOp.getLoc(), newResultTypes, ifOp.getCondition());
+        rewriter.cloneRegionBefore(ifOp.getThenRegion(), newIfOp.getThenRegion(),
+                                   newIfOp.getThenRegion().end());
+        rewriter.cloneRegionBefore(ifOp.getElseRegion(), newIfOp.getElseRegion(),
+                                   newIfOp.getElseRegion().end());
```

Comment:
Cloning can be useful if we need to [remap ssa values](https://mlir.llvm.org/doxygen/classmlir_1_1OpBuilder.html#a79e4f0acdeb3065333057dcac9949404), or need to preserve the original. 
However, in this case it will be more efficient to just [steal the body](https://mlir.llvm.org/doxygen/classmlir_1_1Region.html#ad27658d6e09cba6338419699eabb9c5b) I think:
```suggestion
        newIfOp.getThenRegion().takeBody(ifOp.getThenRegion());
        newIfOp.getElseRegion().takeBody(ifOp.getElseRegion());
```